### PR TITLE
add encryption with age default

### PR DESCRIPTION
**Pull request by MLFlexer**

### CODE REVIEW

File: plugin/init.lua [hunk 1]
```diff
@@ -60,35 +63,82 @@ local function get_file_path(file_name, type, opt_name)
 	return string.format("%s%s" .. separator .. "%s.json", pub.save_state_dir, type, file_name:gsub(separator, "+"))
 end
 
+---@alias encryption_opts {enable: boolean, private_key: string | nil, public_key: string | nil, encrypt: fun(file_path: string, lines: string[]), decrypt: fun(file_path: string): string | nil}
+pub.encryption = {
+	enable = false,
+	private_key = nil,
+	public_key = nil,
+	encrypt = function(file_path, lines)
+		local success, stdout, stderr = wezterm.run_child_process({
```
Comment (MLFlexer) [inline plugin/init.lua:72]: I am not too keen on the arguments to this child process, but I have not managed to find a better way to do this. Please let me know if you know an alternative way of doing this @joncrangle

**Comment by MLFlexer** [comment]
Readme also needs to be updated, will do that later

File: plugin/init.lua (hunk 1 above)
Comment (joncrangle) [reply plugin/init.lua:72]: This implementation has the benefit of keeping data encrypted at rest and only encrypts/decrypts when absolutely needed. Other options I can think of would be less secure because the state would remain decrypted for longer periods of time.

I just submitted a PR to this branch with updates I needed to make to get it working in my config + some doc content.

Comment (MLFlexer) [reply plugin/init.lua:75]: We should stick to that implementation then

File: encryption.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        file_path:gsub(" ", "\\ ")
+      )
+    )
+    if not success then
+      wezterm.log_error(stderr)
+    end
+  end,
+  decrypt = function(file_path)
+    local success, stdout, stderr =
+			execute_shell_cmd(string.format("gpg --decrypt \"%s\" | tr -d '\\000-\\037'", file_path)) -- TODO: fix (\u0000-\u001F) from being included
```
Comment (MLFlexer) [inline encryption.md:138]: There is a problem with chars in the range (\u0000-\u001F) being included in the json, this might not be caused by the decryption, but rather the encryption. Adding `| tr -d '\\000-\\037'` makes it possible to restore some of the state, but it is not possible to restore completely as some of the state has been corrupted by removing the `(\u0000-\u001F)`.

This will need more investigation

Comment (joncrangle) [reply encryption.md:138]: ---@param file_path string
---@param state table
local function write_state(file_path, state)
	local json_state = wezterm.json_encode(state)
	json_state = json_state:gsub("[%c]", "") -- remove control characters
	if pub.encryption.enable then
		pub.encryption.encrypt(file_path, json_state)
	else
		local file = assert(io.open(file_path, "w"))
		file:write(json_state)
		file:close()
	end
end

---@param file_path string
---@return table
local function load_json(file_path)
	if pub.encryption.enable then
		local json = pub.encryption.decrypt(file_path)
		if not json then
			return {}
		end
		json = json:gsub("[%c]", "") -- remove control characters
		return wezterm.json_parse(json)
	else
		local lines = {}
		for line in io.lines(file_path) do
			table.insert(lines, line)
		end
		local json = table.concat(lines)
		return wezterm.json_parse(json)
	end
end

I ended up with this error and was able to avoid it with the above commented lines to sanitize control chars from the json.

Comment (joncrangle) [reply encryption.md:138]: I came across another error with improperly sanitized UTF8 chars.
We may need a `sanitize_json()` function:
```lua
--- Sanitize the input by removing control characters and invalid UTF-8 sequences
--- @param data string
--- @return string
local function sanitize_json(data)
    -- Remove control characters (except for \n, \r, \t)
    data = data:gsub("[%c]", function(c)
        if c == "\n" or c == "\r" or c == "\t" then
            return c
        end
        return ""
    end)
    -- Remove invalid UTF-8 sequences
    return data:gsub("[^\x00-\x7F]", "")
end
```

Comment (MLFlexer) [reply encryption.md:138]: Awesome that you investigated this! and made a PR, really appreciate your contributions :smile: 

Did you figure out the cause of these issues, I did not notice these errors before trying GPG, so i suspect it to be caused by the GPG encryption?

**Comment by MLFlexer** [comment]
I have changed the json sanitizer to replace the invalid escape sequences with what should be valid `\uXXXX` unicode as per the json specification. Hopefully this will fix the issue

**Comment by MLFlexer** [comment]
I think it is ready to be merged now, but I would like to test it a bit more, as there could be edge cases that i have missed

**Comment by joncrangle** [comment]
When I've got some free time I'll test on various OS

**Comment by joncrangle** [comment]
I've tested it and it's working well on Arch and Mac. I've been having a hell of a time trying to get it to work on Windows though.

Using `cmd /c` very quickly breaks down because of character limits, so I've been experimenting with Powershell since it is more capable.
```lua
local process_args = is_windows and { "powershell", "-NonInteractive", "-NoProfile", "-Command", cmd } or
		{ os.getenv("SHELL"), "-c", cmd }
```
I'm currently running into an issue where `wezterm.run_child_process()` is including `Write-Host` content from the `$PROFILE` into the saved state file, making it unusable (even with the above flags) because it's above the actual json content.

Powershell also seems to have character limit issues, so if there's a lot of TTY content it results in a "filepath too long" error and being unable to write the state. I'll give it a more robust go another time. 

One avenue I might pursue is to base64 encode/decode `lines` to avoid some of these issues, and to investigate why the command isn't being echoed as-is when piped to `age`.

**Comment by MLFlexer** [comment]
Thanks for your investigation into the problem, I tried to see if it was possible to ship an buildin encryption provider via. a dynamicly linked library such that we would not need to relie on `run_child_processes`, but from [this comment](https://github.com/wez/wezterm/issues/2088#issuecomment-1148808467) it is clear that it wont be supported.
So i guess the only option is to try and figure out how to overcome the problem on windows

**Comment by MLFlexer** [comment]
I would like to merge this PR and work on the issues for windows in another issue, as the PR has grown quite big. Will add warning to encryption.md and then merge this. The windows issue can continue in #32
